warehouse/query-core/src/main/java/datawave/query/iterator/waitwindow/WaitWindowObserver.java

```diff
@@ -0,0 +1,505 @@
+     */
+    public void start(Range seekRange, long yieldThresholdMs) {
+        this.seekRange = seekRange;
+        this.remainingTimeMs.set(yieldThresholdMs);
+        this.endOfWaitWindow = yieldThresholdMs + System.currentTimeMillis();
+        WaitWindowObserver.getTimer().schedule(this.timerTask, this.checkPeriod, this.checkPeriod);
+    }
+
+    /*
+     * Ensure that the WaitWindowTimerTask is cancelled. Called from QueryIterator.hasTop.
+     */
+    public void stop() {
+        this.timerTask.cancel();
+    }
+
+    /*
+     * Called from waitWindowOverrun() and from places that use a timeout for polling or retrieving Future results remainingTimeMs get updated periodically in
+     * another thread
+     */
+    public long remainingTimeMs() {
+        if (this.yieldCallback == null) {
+            return Long.MAX_VALUE;
+        } else {
+            return this.remainingTimeMs.get();
+        }
+    }
+
+    /*
+     * remainingTimeMs get updated periodically in another thread
+     */
+    public boolean waitWindowOverrun() {
+        return remainingTimeMs.get() <= 0;
+    }
+
+    /*
+     * If we have exceeded the wait window, then immediately throw a WaitWindowOverrunException containing a yield key corresponding to the provided key.
+     */
+    public void checkWaitWindow(Key currentKey, boolean yieldToBeginning) {
+        if (this.yieldCallback != null && waitWindowOverrun()) {
+            Key currentYieldKey = createYieldKey(currentKey, yieldToBeginning);
+            if (this.yieldKey == null) {
+                throw new WaitWindowOverrunException(currentYieldKey);
+            } else {
+                throw new WaitWindowOverrunException(lowestYieldKey(Arrays.asList(this.yieldKey, currentYieldKey)));
+            }
+        }
+    }
+
+    /*
+     * There can be many embedded AndIterators, OrIterators, and Ivarators where a WaitWindowOverrunException can be thrown from. As the exception makes its way
+     * to the top of the call chain, we need to evaluate the yieldKey at each level.
+     */
+    public void propagateException(Key key, boolean yieldToBeginning, boolean keepLowest, WaitWindowOverrunException e) {
+        Key yieldKey;
+        if (key == null) {
+            yieldKey = e.getYieldKey();
+        } else {
+            Collection<Key> keys = Arrays.asList(e.getYieldKey(), createYieldKey(key, yieldToBeginning));
+            if (keepLowest) {
+                yieldKey = lowestYieldKey(keys);
+            } else {
+                yieldKey = highestYieldKey(keys);
+            }
+        }
+        throw new WaitWindowOverrunException(yieldKey);
+    }
+
+    /*
+     * When yieldKey is set and collectTimingDetails=true, then we yield on the second call from QueryIterator.hasTop so that the first call can return the
+     * document that contains the WAIT_WINDOW_OVERRUN and TIMING_METADATA attributes. readyToYield will be set to true after the first call to yieldOnOverrun
```

Comment:
Remind me again, why are we bothering to yield on the second call if we are returning a document with the TIMING_METADATA.  Does not returning a document allow the table to execute the other scans which is the point of yielding?

Reply:
As long as itr.hasTop(), TabletBase continues to add a new KVEntry(key, value) to results and call next().
Documents with TIMING_METADATA and WAIT_WINDOW_OVERRUN are seen as resular documents at this point.
It's necessary to yield after that to force the results back to the caller and rebuild the iterator.